File: training-data/prolog/quicksort.pl
quicksort([], []).
quicksort([Pivot|Rest], Sorted) :-
    partition_list(Pivot, Rest, Smaller, Larger),
    quicksort(Smaller, SortedSmaller),
    quicksort(Larger, SortedLarger),
    append(SortedSmaller, [Pivot|SortedLarger], Sorted).

partition_list(_, [], [], []).
partition_list(Pivot, [H|T], [H|Smaller], Larger) :-
    H < Pivot, !,
    partition_list(Pivot, T, Smaller, Larger).
partition_list(Pivot, [H|T], Smaller, [H|Larger]) :-
    partition_list(Pivot, T, Smaller, Larger).

:- quicksort([5,3,8,1,9,2], Sorted), writeln(Sorted).

File: training-data/ocaml/quicksort.ml
let rec quicksort = function
  | [] -> []
  | pivot :: rest ->
      let smaller = List.filter (fun x -> x < pivot) rest in
      let larger = List.filter (fun x -> x >= pivot) rest in
      quicksort smaller @ [ pivot ] @ quicksort larger

let () =
  quicksort [ 5; 3; 8; 1; 9; 2 ]
  |> List.map string_of_int
  |> String.concat " "
  |> print_endline
